rholang/src/main/scala/coop/rchain/rholang/interpreter/compiler/normalize.scala

```diff
@@ -1061,17 +1191,222 @@ object ProcNormalizeMatcher {
+          than input processes, so that each sequential bind doesn't add a new unforgeable name to the tuplespace.
+          The Rholang 1.1 spec defines them as the latter. Because the Rholang 1.1 spec defines let processes in terms
+          of a output process in concurrent composition with an input process, the let process appears to quote the
+          process on the RHS of "<-" and bind it to the pattern on LHS. For example, in
+
+              let x <- 1 in { Nil }
+
+          the process (value) "1" is quoted and bound to "x" as a name. There is no way to perform an AST transformation
+          of sequential let into a match process and still preserve these semantics, so we have to do an ADT transformation.
+           */
+          case _ =>
+            val newContinuation =
+              p.decls_ match {
+                case _: EmptyDeclImpl => p.proc_
+                case linearDeclsImpl: LinearDeclsImpl =>
+                  val newDecl =
+                    linearDeclsImpl.listlineardecl_.head match {
+                      case impl: LinearDeclImpl => impl.decl_
+                    }
+                  val newDecls =
+                    if (linearDeclsImpl.listlineardecl_.size == 1)
+                      new EmptyDeclImpl()
+                    else {
+                      val newListLinearDecls = new ListLinearDecl()
+                      linearDeclsImpl.listlineardecl_.tail.foreach(newListLinearDecls.add)
+                      new LinearDeclsImpl(newListLinearDecls)
+                    }
+                  new PLet(newDecl, newDecls, p.proc_)
+              }
+
+            def listProcToEList(
+                listProc: List[Proc],
+                knownFree: FreeMap[VarSort]
+            ): M[ProcVisitOutputs] =
+              listProc
+                .foldM((Vector.empty[Par], knownFree, BitSet.empty, false)) {
+                  case ((vectorPar, knownFree, locallyFree, connectiveUsed), proc) =>
+                    ProcNormalizeMatcher
+                      .normalizeMatch[M](
+                        proc,
+                        ProcVisitInputs(VectorPar(), input.boundMapChain, knownFree)
+                      )
+                      .map {
+                        case ProcVisitOutputs(par, updatedKnownFree) =>
+                          (
+                            par +: vectorPar,
+                            updatedKnownFree,
+                            locallyFree | par.locallyFree,
+                            connectiveUsed | par.connectiveUsed
+                          )
+                      }
+                }
+                .map {
+                  case (vectorPar, knownFree, locallyFree, connectiveUsed) =>
+                    ProcVisitOutputs(
+                      EList(vectorPar.reverse, locallyFree, connectiveUsed, none[Var]),
+                      knownFree
+                    )
+                }
+
+            // Largely similar to how consume patterns are processed.
```

Comment:
It would be better to write a re-usable function for similar process

Reply:
Agreed, but that's not feasible without significant compiler refactoring.

Reply:
@zsluedem I created an issue for refactoring of normalizer #3290.